### PR TITLE
Layers view shortcut fixes

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Description of the Change
This change fixes issues with the following shortcuts.
Deselect
New Layer
Enable Layer

All of the above shortcuts now work as intended with the view opened, as well as the view closed.

Also added during this PR, is an accessor method to determine whether the TopComponent is visible or not. This was used to determine if a manual update of queries was necessary. (When the component is open, the state changed attribute handler usually takes care of the updating of queries.)

### Alternate Designs
None

### Why Should This Be In Core?
Shortcuts now work

### Benefits
Shortcuts work as intended

### Verification Process
Opening new and saved graphs for each of the below actions.

- Make new layers with LV opened, Make new layers with LV closed
- Deselect all layers with/Without LV opened
- Enable layers with and without the LV opened.

Problems which kept this PR coming through was the fact that it worked perfect with the view opened, as it updated the view using the attribute change handlers. When the view was not opened, it would not update. A manual update would cause issues with the operation when the LV was opened. An if case was added to test if the view was opened, and a manual update was performed from that.

### Applicable Issues
#823 #725

### CODE REVIEW

File: CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/expression/ExpressionParser.java [hunk 1]
```diff
@@ -323,7 +322,7 @@ protected void print(String prefix, StringBuilder out) {
     public static SequenceExpression parse(String expression) {
 
         ParseState state = ParseState.READING_WHITESPACE;
-        final char[] content = new char[expression.length()];
+        char[] content = new char[expression.length()];
```
Comment (antares1470) [inline CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/expression/ExpressionParser.java:325]: final

File: CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/operations/And.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
  */
 public class And {
 
-    public static final String NAME = "AND";
+    public static final String NAME = new String("AND");
```
Comment (antares1470) [inline CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/operations/And.java:28]: What's the reason for using the constructor here (and other related files)?

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 3]
```diff
@@ -147,6 +151,22 @@ public void execute() {
                 .executeLater(GraphManager.getDefault().getActiveGraph());
     }
 
+    public void executeFuture() {
+        final int newBitmask = LayersUtilities.calculateCurrentLayerSelectionBitMask(vxBitMaskCollection, txBitMaskCollection);
+
+        Future<?> f = PluginExecution.withPlugin(new UpdateLayerSelectionPlugin(newBitmask))
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:157]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 4]
```diff
@@ -147,6 +151,22 @@ public void execute() {
...
+                .executeLater(GraphManager.getDefault().getActiveGraph());
+
+        try {
+            f.get();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:165]: exceptions can be final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 5]
```diff
@@ -161,6 +181,28 @@ public void readState() {
...
+        return parent != null ? parent.getVisibility() : null;
+    }
+
+    public void readStateFuture() {
+        final LayersViewPane pane = parent.getContent();
+        final Graph graph = GraphManager.getDefault().getActiveGraph();
+        if (pane == null || graph == null) {
+            return;
+        }
+        Future<?> f = PluginExecution.withPlugin(new LayersViewStateReader(pane))
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:194]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 6]
```diff
@@ -161,6 +181,28 @@ public void readState() {
...
+                .executeLater(graph);
+        try {
+            f.get();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:201]: both exceptions final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 7]
```diff
@@ -172,13 +214,36 @@ public void readState() {
...
+    }
+
+    public void updateQueriesFuture(final Graph currentGraph) {
+        final Graph graph = GraphManager.getDefault().getActiveGraph();
+        if (graph == null) {
+            return;
+        }
         final UpdateQueryPlugin updatePlugin = new UpdateQueryPlugin(vxBitMaskCollection, txBitMaskCollection);
-        PluginExecution.withPlugin(updatePlugin).executeLater(currentGraph);
+        Future<?> f = PluginExecution.withPlugin(updatePlugin).executeLater(graph);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:238]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 8]
```diff
@@ -172,13 +214,36 @@ public void readState() {
...
+        try {
+            f.get();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:244]: both exceptions final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java [hunk 9]
```diff
@@ -52,74 +54,87 @@
 @NbBundle.Messages("CTL_LayersViewShortcuts=Layers View: Shortcuts")
 public class LayersViewShortcuts extends AbstractAction {
 
+    private void triggerLayerSelection(final int layerId) {
+        var vxCollection = LayersViewController.getDefault().getVxQueryCollection();
+        var txCollection = LayersViewController.getDefault().getTxQueryCollection();
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java:59]: is there an alternative to declaring `var`?

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java [hunk 10]
```diff
@@ -52,74 +54,87 @@
...
+        if (txCollection.getQuery(layerId) != null) {
+            txCollection.getQuery(layerId).setVisibility(!txCollection.getQuery(layerId).getVisibility());
+        }
+
+        final int newBitmask = LayersUtilities.calculateCurrentLayerSelectionBitMask(vxCollection, txCollection);
+
+        PluginExecution.withPlugin(new UpdateLayerSelectionPlugin(newBitmask))
+                .executeLater(GraphManager.getDefault().getActiveGraph());
+
+        Future<?> future = LayersViewController.getDefault().writeState();
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java:73]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java [hunk 11]
```diff
@@ -52,74 +54,87 @@
...
+
+        try {
+            future.get();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java:80]: both exceptions final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java [hunk 12]
```diff
@@ -52,74 +54,87 @@
...
-        Future<?> enableLayerFuture = null;
-
         final String hotkey = e.getActionCommand();
         switch (hotkey) {
             case "CA-L":
                 PluginExecution.withPlugin(new NewLayerPlugin()).executeLater(GraphManager.getDefault().getActiveGraph());
                 break;
             case "CA-D":
-                PluginExecution.withPlugin(new DeselectAllLayersPlugin()).executeLater(GraphManager.getDefault().getActiveGraph());
+                Future<?> deselectFuture = PluginExecution.withPlugin(new DeselectAllLayersPlugin()).executeLater(GraphManager.getDefault().getActiveGraph());
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java:98]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java [hunk 13]
```diff
@@ -52,74 +54,87 @@
...
+                try {
+                    deselectFuture.get();
+                } catch (InterruptedException ex) {
+                    Exceptions.printStackTrace(ex);
+                    Thread.currentThread().interrupt();
+                } catch (ExecutionException ex) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java:104]: both exceptions final

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/operations/And.java (hunk 2 above)
Comment (aldebaran30701) [reply CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/operations/And.java:28]: So for some reason, the map structure that this gets stored into, does not work when you omit the new String(...) constructor. I previously fixed this issue up in the SQ warnings from a few weeks back - but this specifically broke the Layers View. It was pretty hard to debug.
I've reverted it back which fixed the issue.

Comment (antares1470) [reply CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/value/operations/And.java:28]: hmmm that's really odd. Perhaps the constructor actually behaves differently than expected (though I don't understand why)

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java (hunk 9 above)
Comment (aldebaran30701) [reply CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/shortcut/LayersViewShortcuts.java]: Forgot to remove this var when using it to test my thinking 👍

**Comment by aldebaran30701** [comment]
@antares1470 Should have all those comments fixed. 👍

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**